### PR TITLE
Voice search support

### LINKED ISSUE #405: Implement Voice Search in FxR for 1.0
## Hardware

All

## Steps to Reproduce

1. NA

## Current Behavior

Selecting microphone button in URL bar does nothing.

## Expected Behavior

Selecting microphone should allow user to do a voice search that results in a Google search results page. (Note: V1.0 of browser search defaults to Google only, we will implement search engine personalization in 1.1+)

User Flow attached, please ping @thenadj for questions.
[UF-17 Voice Command_002.pdf](https://github.com/MozillaReality/FirefoxReality/files/2315231/UF-17.Voice.Command_002.pdf)

```
```

</details>

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #405 Added support for voice search thatnks to @andrenatal's work on [Mozilla Speech Service](https://github.com/mozilla/androidspeech).

### CODE REVIEW

**Comment by bluemarvin** [comment]
I get a build failure:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all files for configuration ':app:googlevrDebugCompileClasspath'.
> Could not find com.github.mozilla:mozillaspeechlibrary:1.0.3.
  Searched in the following locations:
      file:/Users/rbarker/Library/Android/sdk/extras/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      file:/Users/rbarker/Library/Android/sdk/extras/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      file:/Users/rbarker/Library/Android/sdk/extras/google/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      file:/Users/rbarker/Library/Android/sdk/extras/google/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      file:/Users/rbarker/Library/Android/sdk/extras/android/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      file:/Users/rbarker/Library/Android/sdk/extras/android/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
```

**Comment by keianhzo** [comment]
@bluemarvin I think we may need to wait a bit until that AAR is available in Maven @andrenatal has already pushed it but it may take some time.

**Comment by bluemarvin** [comment]
What maven repository is it getting pushed to? This is where we currently search:

```
allprojects {
    repositories {
        google()
        jcenter()

        maven {
            url 'https://maven.mozilla.org/maven2'
        }
    }
}
```

**Comment by andrenatal** [comment]
@bluemarvin sonatype's https://search.maven.org/

**Comment by bluemarvin** [comment]
Has it been uploaded, I added sonatype:

```
allprojects {
    repositories {
        google()
        jcenter()

        maven {
            url 'https://maven.mozilla.org/maven2'
        }
        maven {
            url 'https://search.maven.org/'
        }
    }
}
```
and still get:

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all files for configuration ':app:googlevrDebugCompileClasspath'.
> Could not find com.github.mozilla:mozillaspeechlibrary:1.0.3.
  Searched in the following locations:
      file:/Users/rbarker/Library/Android/sdk/extras/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      file:/Users/rbarker/Library/Android/sdk/extras/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      file:/Users/rbarker/Library/Android/sdk/extras/google/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      file:/Users/rbarker/Library/Android/sdk/extras/google/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      file:/Users/rbarker/Library/Android/sdk/extras/android/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      file:/Users/rbarker/Library/Android/sdk/extras/android/m2repository/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      https://dl.google.com/dl/android/maven2/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      https://dl.google.com/dl/android/maven2/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      https://jcenter.bintray.com/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      https://jcenter.bintray.com/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      https://maven.mozilla.org/maven2/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      https://maven.mozilla.org/maven2/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
      https://search.maven.org/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.pom
      https://search.maven.org/com/github/mozilla/mozillaspeechlibrary/1.0.3/mozillaspeechlibrary-1.0.3.jar
  Required by:
      project :app
```

**Comment by andrenatal** [comment]
@bluemarvin please use to test. The library is still being propagated and will be available under google's maven soon

```        
        maven {
            url "https://oss.sonatype.org/content/repositories/releases/"
        }

```

**Comment by andrenatal** [comment]
I just tested here together @nancyhang and @kgiori and is working! Very cool interface btw! Will review the code.

**Comment by bluemarvin** [comment]
I get frequent crashes an application freezes when doing voice search on my Pixel 2. It usually only takes four or five searches to crash.

**Comment by bluemarvin** [comment]
I get frequent crashes on the Go too. From the looks of the stack traces I would say it is leaking memory until we run out.

**Comment by andrenatal** [comment]
I didn't get any crashes on the Go. What you did to get that?

**Comment by andrenatal** [comment]
Just tested and seems the crashes are gone both on Pixel as on the Go. Could you guys please test it?

In the phone the permission dialog that got displayed was the default for android though, consequently was impossible to see and accept it while wearing the daydream. Needed to remove the phone from the headset and tap on accept.

File: app/src/common/shared/org/mozilla/vrbrowser/ui/NavigationBarWidget.java [hunk 1]
```diff
@@ -589,4 +597,32 @@ public void onCurrentSessionChange(GeckoSession aSession, int aId) {
             button.setPrivateMode(isPrivateMode);
         }
     }
+
+    @Override
+    public void OnVoiceSearchClicked() {
+        VoiceSearchWidget widget = getChild(mVoiceSearchInputHandle);
+        if (widget == null) {
+            widget = createChild(VoiceSearchWidget.class, false);
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/ui/NavigationBarWidget.java]: Fixed

File: app/build.gradle [hunk 2]
```diff
@@ -188,6 +188,7 @@ dependencies {
     implementation 'com.android.support:design:27.1.1'
     implementation 'com.google.vr:sdk-audio:1.160.0'
     implementation "org.mozilla.components:telemetry:0.10"
+    implementation "com.github.mozilla:mozillaspeechlibrary:1.0.3"
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply app/build.gradle]: Updated

File: app/src/common/shared/org/mozilla/vrbrowser/ui/VoiceSearchWidget.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+                        case STT_RESULT:
+                            // When the api finished processing and returned a hypothesis
+                            Log.d(LOGTAG, "===> STT_RESULT");
+                            String transcription = ((STTResult)aPayload).mTranscription;
+                            float confidence = ((STTResult)aPayload).mConfidence;
+                            if (mDelegate != null)
+                                mDelegate.OnVoiceSearchResult(transcription, confidence);
+                            hide();
+                            break;
+                        case START_LISTEN:
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply app/src/common/shared/org/mozilla/vrbrowser/ui/VoiceSearchWidget.java]: I've debugged it and it seems that after a recognition is started (after an error) the first status change callback that we get is an error state. 

As the listener is attached to the MozillaSpeechService singleton when we attach the listener after a subsequent session we still get callbacks from the previous session.

**Comment by bluemarvin** [comment]
@andrenatal Having to take the phone out of the head set for Pixel is sadly how it works. On all-in-one Daydream headsets like the Mirage it obviously works fine. I have no idea why Google uses such an awkward permission flow on phone based daydream devices.

**Comment by keianhzo** [comment]
@bluemarvin It seems that Google defines a flow for transitioning to 2D monoscopic. I guess we will need to address this before publishing for Daydream.

> - Prompt the user to remove their phone.
> - Detect the orientation change to portrait to switch over to 2D.
> - Return the user to nearly the same state in VR that they were in before they removed their phone.
> 
> If an app requires an Android permission, the permission must follow [UX-D9](https://developers.google.com/vr/distribute/daydream/design-requirements#UX-D9) to accept the permission. See this page on the [Video360 sample](https://developers.google.com/vr/android/samples/video360#permissions_and_hybrid_vr_app_flow) for more information on accepting permissions.

**Comment by bluemarvin** [comment]
@keianhzo Are we shipping on Daydream clip-in phones? Last I heard we weren't planning on it.

**Comment by nancyhang** [comment]
We're not shipping on clip-in Daydream headsets @keianhzo @bluemarvin

**Comment by keianhzo** [comment]
Added the permission denied dialog. 

@andrenatal @bluemarvin I think the only missing thing is the error message after a failed search (only happens if you start a new search quick enough). @andrenatal is looking into it but I don't see it as a blocker so if everything looks good I'd rather land this so people can try it as soon as possible.

**Pull request merged**